src/Processors/Formats/Impl/FreeformRowInputFormat.cpp

```diff
@@ -0,0 +1,473 @@
+}
+
+std::vector<std::pair<String, String>> JSONFieldMatcher::parseFields(ReadBuffer & in, size_t index) const
+{
+    if (in.eof())
+        return {};
+
+    if (*in.position() != '{')
+    {
+        String ret;
+        readJSONField(ret, in);
+        return {{fmt::format("c{}", index), ret}};
+    }
+
+
+    ++in.position();
+    skipWhitespacesAndDelimiters(in);
+
+    std::vector<std::pair<String, String>> ret;
+    while (*in.position() != '}')
+    {
+        String field_name = JSONUtils::readFieldName(in);
+
+        auto fields = parseFields(in, index); // "index" here doesn't do anything
+
+        for (const auto & [name, field] : fields)
+        {
+            if (name.starts_with("c"))
+                ret.emplace_back(field_name, field);
+            else
+                ret.emplace_back(fmt::format("{}.{}", field_name, name), field);
+        }
+
+        skipWhitespacesAndDelimiters(in);
+    }
+
+    ++in.position();
+
+    return ret;
+}
+
+std::vector<std::pair<String, String>> CSVFieldMatcher::parseFields(ReadBuffer & in, size_t index) const
+{
+    String ret;
+    readCSVField(ret, in, settings.csv);
+    return {{fmt::format("c{}", index), ret}};
+}
+
+std::vector<std::pair<String, String>> QuotedFieldMatcher::parseFields(ReadBuffer & in, size_t index) const
+{
+    String ret;
+    readQuotedField(ret, in);
+    return {{fmt::format("c{}", index), ret}};
+}
+
+std::vector<std::pair<String, String>> EscapedFieldMatcher::parseFields(ReadBuffer & in, size_t index) const
+{
+    String ret;
+    readEscapedString(ret, in);
+    return {{fmt::format("c{}", index), ret}};
+}
+
+std::vector<std::pair<String, String>> RawByWhitespaceFieldMatcher::parseFields(ReadBuffer & in, size_t index) const
+{
+    String ret;
+    readStringUntilWhitespaceDelimiter(ret, in);
+    return {{fmt::format("c{}", index), ret}};
+}
+
+FreeformFieldMatcher::FreeformFieldMatcher(ReadBuffer & in_, const FormatSettings & settings_)
+    : max_rows_to_check(std::min<size_t>(100, settings_.max_rows_to_read_for_schema_inference)), in(in_)
+{
+    // matchers are pushed in the order of priority, this helps with exiting early and reducing the search tree.
+    matchers.emplace_back(std::make_unique<JSONFieldMatcher>(FormatSettings::EscapingRule::JSON, settings_));
+    matchers.emplace_back(std::make_unique<CSVFieldMatcher>(FormatSettings::EscapingRule::CSV, settings_));
+    matchers.emplace_back(std::make_unique<RawByWhitespaceFieldMatcher>(FormatSettings::EscapingRule::Raw, settings_));
+    matchers.emplace_back(std::make_unique<QuotedFieldMatcher>(FormatSettings::EscapingRule::Quoted, settings_));
+    matchers.emplace_back(std::make_unique<EscapedFieldMatcher>(FormatSettings::EscapingRule::Escaped, settings_));
+}
+
+std::vector<FreeformFieldMatcher::Fields> FreeformFieldMatcher::readNextFields(bool one_string, size_t index)
+{
+    skipWhitespacesAndDelimiters(in);
+    char * start = in.position();
```

Comment:
I think you don't fully understand how our ReadBuffers works.
ReadBuffer contains a buffer with limited size, when current data is fully read from the buffer, we fill this buffer with the new data and start reading from the start of the same buffer (the end of buffer can change depending on how much new data we read). So, we cannot just remember current position of the buffer and then get back to it just by remembering in.position(), because the data in the buffer can change and we will rollback to invalid position. It can cause segfaults and unexpected behaviour.
I see that you use it a lot in this PR when we need to get back to some previous position of the buffer. There are two options how to deal with it:
1) We have special ReadBuffer for such cases: PeekableReadBuffer. It allows to make checkpoints in the data and then rollback to it. Checkpoints can be recursive.
2) Read each row of input data in a String and then use ReadBufferFromString. In this case we will be sure that the data and the buffer won't change.
I prefer the first solution. If you will have any questions about PeekableReadBuffer, you can ask me.

Reply:
I see, I will try to play around with `PeakableReadBuffer`, based on what you said it will give the desirable behavior